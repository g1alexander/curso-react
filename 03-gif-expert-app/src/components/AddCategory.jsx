import { useState } from "react";
import PropTypes from "prop-types";

export function AddCategory({ setCategories }) {
  const [inputValue, setInputValue] = useState("");

  const handleInputChange = (e) => {
    setInputValue(e.target.value);
  };

  const handleSubmit = (e) => {
    e.preventDefault();
    if (inputValue.trim().length > 2) {
      setCategories((categories) => [inputValue, ...categories]);
      setInputValue("");
    }
  };

  return (
    <form onSubmit={handleSubmit}>
      <p>{inputValue}</p>
      <input
        value={inputValue}
        onChange={handleInputChange}
        type="text"
        name="add"
        id="add"
      />
    </form>
  );
}

AddCategory.propTypes = {
  setCategories: PropTypes.func.isRequired,
};
