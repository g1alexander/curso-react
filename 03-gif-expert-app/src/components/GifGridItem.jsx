import PropTypes from "prop-types";

export function GifGridItem({ title, image }) {
  return (
    <div className="card animate__animated animate__bounce">
      <h3>{title}</h3>
      <img src={image} alt={title} />
    </div>
  );
}

GifGridItem.propTypes = {
  title: PropTypes.string.isRequired,
  image: PropTypes.string.isRequired,
};
