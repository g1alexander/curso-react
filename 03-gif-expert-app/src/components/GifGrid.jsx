import PropTypes from "prop-types";
import { GifGridItem } from "./GifGridItem";
import { useFetchGif } from "../hooks/useFetchGif";

export function GifGrid({ category }) {
  const { data: images, loading } = useFetchGif(category);

  return (
    <>
      <h2>{category}</h2>

      {loading && (
        <p className="animate__animated animate__flash">cargando...</p>
      )}

      <section className="grid">
        {images.map((img) => (
          <GifGridItem key={img.id} {...img} />
        ))}
      </section>
    </>
  );
}

GifGrid.propTypes = {
  category: PropTypes.string.isRequired,
};
