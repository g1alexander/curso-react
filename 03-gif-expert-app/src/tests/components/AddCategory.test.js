import { shallow } from "enzyme";
import { AddCategory } from "../../components/AddCategory";

describe("pruebas del componente <AddCategory />", () => {
  const setCategories = jest.fn();
  let wrapper = shallow(<AddCategory setCategories={setCategories} />);

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategories={setCategories} />);
  });

  test("debe mostrarse correctamente", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("debe de cambiar la caja de texto", () => {
    const $input = wrapper.find("input"),
      value = "hola mundo";

    $input.simulate("change", { target: { value } });

    expect(wrapper.find("p").text().trim()).toBe(value);
  });

  test("NO debe postear la informacion con submit", () => {
    wrapper.find("form").simulate("submit", { preventDefault() {} });

    expect(setCategories).not.toHaveBeenCalled();
  });

  test("debe de llamar el setCategories y limpiar la caja de texto", () => {
    const $input = wrapper.find("input"),
      value = "hello there";

    $input.simulate("change", { target: { value } });
    wrapper.find("form").simulate("submit", { preventDefault() {} });

    // expect(setCategories).toHaveBeenCalled();
    expect(setCategories).toHaveBeenCalledTimes(1);
    expect(setCategories).toHaveBeenCalledWith(expect.any(Function));
    expect($input.prop("value")).toBe("");
  });
});
