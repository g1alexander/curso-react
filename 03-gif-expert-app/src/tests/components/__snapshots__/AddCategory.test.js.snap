// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas del componente <AddCategory /> debe mostrarse correctamente 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    id="add"
    name="add"
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
